File: packages/coding-agent/src/core/memory/project.ts
import { execFileSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync, realpathSync } from "node:fs";
import { join, resolve } from "node:path";
import { lockSync } from "proper-lockfile";
import { writeFileAtomicSync } from "../../utils/atomic-file.js";
import { hash } from "./evidence.js";

export interface ProjectIdentity {
	id: string;
	root: string;
	aliases: string[];
}
export function normalizeRemote(remote: string): string {
	const value = remote
		.trim()
		.replace(/\.git\/?$/, "")
		.replace(/\/$/, "");
	const scp = value.match(/^(?:[^/@]+@)?([^/:]+):([^/].*)$/);
	if (scp && !value.includes("://") && !/^[A-Za-z]:/.test(value)) return `${scp[1].toLowerCase()}/${scp[2]}`;
	try {
		const url = new URL(value);
		return `${url.hostname.toLowerCase()}${url.port ? `:${url.port}` : ""}${url.pathname}`;
	} catch {
		return value;
	}
}
function git(cwd: string, args: string[]): string | undefined {
	try {
		return (
			execFileSync("git", ["-C", cwd, ...args], {
				encoding: "utf8",
				timeout: 1500,
				stdio: ["ignore", "pipe", "ignore"],
			}).trim() || undefined
		);
	} catch {
		return undefined;
	}
}
export function projectIdentity(cwd: string, agentDir: string, bindId?: string): ProjectIdentity {
	const root = git(cwd, ["rev-parse", "--show-toplevel"]) ?? realpathSync(cwd);
	const common = git(cwd, ["rev-parse", "--path-format=absolute", "--git-common-dir"]) ?? root;
	const remote = git(cwd, ["config", "--get", "remote.origin.url"]);
	const aliases = [`path:${resolve(common)}`, ...(remote ? [`remote:${normalizeRemote(remote)}`] : [])];
	const dir = join(agentDir, "memory");
	mkdirSync(dir, { recursive: true, mode: 0o700 });
	const file = join(dir, "projects.json");
	const release = lockSync(dir, { realpath: true });
	try {
		const registry: Record<string, string> = existsSync(file) ? JSON.parse(readFileSync(file, "utf8")) : {};
		const existing = aliases.map((alias) => registry[alias]).filter(Boolean);
		const id = bindId ?? existing[0] ?? `project_${hash(aliases.at(-1)!).slice(0, 24)}`;
		if (!/^project_[a-zA-Z0-9_-]{1,80}$/.test(id))
			throw new Error("Project ID must start with project_ and contain only letters, digits, _ or -");
		if (!bindId && existing.some((value) => value !== id))
			throw new Error("Project aliases conflict; explicitly bind the intended project ID");
		if (aliases.some((alias) => registry[alias] !== id)) {
			for (const alias of aliases) registry[alias] = id;
			writeFileAtomicSync(file, JSON.stringify(registry, null, 2), { mode: 0o600 });
		}
		return { id, root, aliases };
	} finally {
		release();
	}
}

File: crates/pi-coding-agent/src/core/memory/project.rs
//! Port of packages/coding-agent/src/core/memory/project.ts
use std::path::{Path, PathBuf};
use std::process::Command;

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::utils::atomic_file::write_file_atomic_sync;

use super::evidence::hash;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProjectIdentity {
	pub id: String,
	pub root: String,
	pub aliases: Vec<String>,
}

pub fn normalize_remote(remote: &str) -> String {
	let value = {
		let trimmed = remote.trim();
		let value = regex::Regex::new(r"\.git/?$").unwrap().replace(trimmed, "").to_string();
		regex::Regex::new(r"/$").unwrap().replace(&value, "").to_string()
	};
	let scp_re = regex::Regex::new(r"^(?:[^/@]+@)?([^/:]+):([^/].*)$").unwrap();
	let drive_re = regex::Regex::new(r"^[A-Za-z]:").unwrap();
	if let Some(captures) = scp_re.captures(&value) {
		if !value.contains("://") && !drive_re.is_match(&value) {
			let host = captures.get(1).map(|m| m.as_str()).unwrap_or("");
			let path = captures.get(2).map(|m| m.as_str()).unwrap_or("");
			return format!("{}/{}", host.to_lowercase(), path);
		}
	}
	match url::Url::parse(&value) {
		Ok(url) => {
			let hostname = url.host_str().unwrap_or("").to_lowercase();
			let port = url.port().map(|port| format!(":{port}")).unwrap_or_default();
			format!("{hostname}{port}{}", url.path())
		}
		Err(_) => value,
	}
}

fn git(cwd: &str, args: &[&str]) -> Option<String> {
	let output = Command::new("git")
		.arg("-C")
		.arg(cwd)
		.args(args)
		.output()
		.ok()?;
	if !output.status.success() {
		return None;
	}
	let trimmed = String::from_utf8_lossy(&output.stdout).trim().to_string();
	if trimmed.is_empty() {
		None
	} else {
		Some(trimmed)
	}
}

fn realpath_sync(path: &Path) -> String {
	std::fs::canonicalize(path)
		.map(|value| value.to_string_lossy().to_string())
		.unwrap_or_else(|_| path.to_string_lossy().to_string())
}

/// Windows paths are case-insensitive; lock identity must match the OS view.
fn lock_key(path: &Path) -> String {
	realpath_sync(path).to_lowercase()
}

fn projects_registry_path(agent_dir: &str) -> PathBuf {
	Path::new(agent_dir).join("memory").join("projects.json")
}

fn read_registry(file: &Path) -> IndexMap<String, Value> {
	match std::fs::read_to_string(file) {
		Ok(raw) => match serde_json::from_str::<Value>(&raw) {
			Ok(Value::Object(map)) => map.into_iter().collect(),
			_ => IndexMap::new(),
		},
		Err(_) => IndexMap::new(),
	}
}

pub fn project_identity(cwd: &str, agent_dir: &str, bind_id: Option<&str>) -> Result<ProjectIdentity, String> {
	let root = git(cwd, &["rev-parse", "--show-toplevel"]).unwrap_or_else(|| realpath_sync(Path::new(cwd)));
	let common = git(cwd, &["rev-parse", "--path-format=absolute", "--git-common-dir"]).unwrap_or_else(|| root.clone());
	let remote = git(cwd, &["config", "--get", "remote.origin.url"]);
	let mut aliases = vec![format!("path:{}", realpath_sync(Path::new(&common)))];
	if let Some(remote) = &remote {
		aliases.push(format!("remote:{}", normalize_remote(remote)));
	}
	let dir = Path::new(agent_dir).join("memory");
	std::fs::create_dir_all(&dir).map_err(|error| error.to_string())?;
	let file = dir.join("projects.json");
	let _lock = crate::utils::dir_lock::lock_dir_sync(&lock_key(&dir), None);
	let registry = read_registry(&file);
	let existing: Vec<String> = aliases
		.iter()
		.filter_map(|alias| registry.get(alias))
		.filter_map(Value::as_str)
		.map(str::to_string)
		.collect();
	let id = match bind_id {
		Some(value) => value.to_string(),
		None => match existing.first() {
			Some(value) => value.clone(),
			None => {
				let last = aliases.last().cloned().unwrap_or_default();
				let digest = hash(&last);
				format!("project_{}", &digest[..24.min(digest.len())])
			}
		},
	};
	let id_re = regex::Regex::new(r"^project_[a-zA-Z0-9_-]{1,80}$").unwrap();
	if !id_re.is_match(&id) {
		return Err("Project ID must start with project_ and contain only letters, digits, _ or -".to_string());
	}
	if bind_id.is_none() && existing.iter().any(|value| value != &id) {
		return Err("Project aliases conflict; explicitly bind the intended project ID".to_string());
	}
	if aliases.iter().any(|alias| registry.get(alias).and_then(Value::as_str) != Some(id.as_str())) {
		let mut next: Map<String, Value> = Map::new();
		for (key, value) in &registry {
			next.insert(key.clone(), value.clone());
		}
		for alias in &aliases {
			next.insert(alias.clone(), Value::String(id.clone()));
		}
		let body = format!("{}\n", serde_json::to_string_pretty(&Value::Object(next)).unwrap_or_default());
		write_file_atomic_sync(
			&file.to_string_lossy(),
			&body,
			crate::utils::atomic_file::WriteFileAtomicOptions {
				mode: Some(0o600),
				..Default::default()
			},
		)
		.map_err(|error| error.to_string())?;
	}
	Ok(ProjectIdentity { id, root, aliases })
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn normalizes_authenticated_urls_without_leaking_credentials() {
		assert_eq!(
			normalize_remote("https://secret@github.com/telemusai/optimus-agent.git"),
			normalize_remote("git@github.com:telemusai/optimus-agent.git")
		);
		assert_eq!(normalize_remote("https://github.com/telemusai/prime.git"), "github.com/telemusai/prime");
		assert_eq!(normalize_remote("git@github.com:telemusai/optimus-agent.git"), "github.com/telemusai/optimus-agent");
	}

	#[test]
	fn keeps_ports_and_local_paths_intact() {
		assert_eq!(normalize_remote("https://example.com:8443/x/y/"), "example.com:8443/x/y");
		assert_eq!(normalize_remote("C:/Users/x/repo"), "C:/Users/x/repo");
	}

	#[test]
	fn retains_identity_through_remote_rename_and_honours_explicit_bind() {
		let base = std::env::temp_dir().join(format!("prime-project-{}", uuid::Uuid::new_v4()));
		let cwd = base.join("repo");
		let agent_dir = base.join("agent");
		std::fs::create_dir_all(&cwd).unwrap();
		std::fs::create_dir_all(&agent_dir).unwrap();
		let cwd = cwd.to_string_lossy().to_string();
		let agent_dir = agent_dir.to_string_lossy().to_string();
		let status = Command::new("git").arg("init").arg(&cwd).output().unwrap();
		assert!(status.status.success());
		let _ = Command::new("git")
			.args(["-C", &cwd, "remote", "add", "origin", "https://github.com/telemusai/prime.git"])
			.output()
			.unwrap();
		let first = project_identity(&cwd, &agent_dir, None).expect("identity");
		let _ = Command::new("git")
			.args([
				"-C",
				&cwd,
				"remote",
				"set-url",
				"origin",
				"git@github.com:telemusai/optimus-agent.git",
			])
			.output()
			.unwrap();
		assert_eq!(project_identity(&cwd, &agent_dir, None).expect("identity").id, first.id);
		assert_eq!(
			project_identity(&cwd, &agent_dir, Some("project_explicit"))
				.expect("identity")
				.id,
			"project_explicit"
		);
		std::fs::remove_dir_all(&base).ok();
	}
}
